crates/re_renderer/src/allocator/data_texture_source.rs

```diff
@@ -0,0 +1,317 @@
+    ///      -> Con: We waste space and copying to the texture is harder
+    ///
+    /// We're going with option (1)!
+    ///
+    /// This means that there might be 'n' full buffers followed by a single active buffer, followed by 'm' empty buffers.
+    buffers: Vec<CpuWriteGpuReadBuffer<T>>,
+
+    /// Buffer in which data is currently written.
+    ///
+    /// All buffers before are full and all after (if any) are empty.
+    ///
+    /// After `reserve` it is guaranteed to point to a buffer with remaining capacity.
+    /// Prior to that, it may point to no buffer at all or a full buffer.
+    active_buffer_index: usize,
+}
+
+impl<'a, T: Pod + Send + Sync> DataTextureSource<'a, T> {
+    /// Creates a new `DataTextureSource` with the given `RenderContext`.
+    ///
+    /// This operation itself will not allocate any memory, empty `DataTextureSource` are not a concern.
+    pub fn new(ctx: &'a RenderContext) -> Self {
+        Self {
+            ctx,
+            buffers: Vec::new(),
+            active_buffer_index: 0,
+        }
+    }
+
+    #[inline]
+    fn max_data_texture_width(&self) -> usize {
+        // We limit the data texture width to 16384 or whatever smaller value is supported but the device.
+        //
+        // If we make buffers always a multiple of this width, we can do all buffer copies in a single copy!
+        //
+        // But wait! If we're using this as the minimum buffer size, isn't that too big?
+        // 16384 * float4 (worst case) = 256KiB.
+        // Keep in mind that weaker hardware will have 8192 max width.
+        // Also note, that many of our textures use 4 & 8 byte formats.
+        // So while this is still a considerable amount of memory when used for very small data textures
+        // it's not as bad as it seems.
+        // Given how much it simplifies to keep buffers a multiple of the texture width,
+        // this seems to be a reasonable trade-off.
+        (self.ctx.device.limits().max_texture_dimension_2d as usize).min(16384)
+    }
+
+    /// The number of elements written so far.
+    #[inline]
+    pub fn num_written(&self) -> usize {
+        self.buffers.iter().map(|b| b.num_written()).sum()
+    }
+
+    /// Reserves space for at least `num_elements` elements.
+    ///
+    /// Creating new buffers is a relatively expensive operation, so it's best to
+    /// reserve gratuitously!
+    /// Ideally, you know exactly how many elements you're going to write and reserve
+    /// accordingly at the start.
+    pub fn reserve(&mut self, num_elements: usize) -> Result<(), CpuWriteGpuReadError> {
+        let remaining_capacity: usize = self.buffers.iter().map(|b| b.remaining_capacity()).sum();
+        if remaining_capacity >= num_elements {
+            return Ok(());
+        }
+
+        // Constraints on the buffer size:
+        // * have at least as many elements as requested
+        // * be at least double the size of the last buffer
```

Comment:
I think this is a great allocation strategy, and it should be documented in the docstring!

In particular, I don't want to worry users that `reserve` is allocating _too little_. A user should feel free to call `reserve` often, with small increments if the situation calls for it